chipsec/chipset.py

```diff
@@ -912,24 +927,59 @@ def write_register(self, reg_name, reg_value, cpu_thread=0, bus_index=0):
-            return True
+        reg = self.get_register_def(reg_name)
+        rtype = reg['type']
+        bus_data = self.get_register_bus(reg_name)
+        ret = False
+        if RegisterType.MSR == rtype:
+            topology = self.cpu.get_cpu_topology()
+            if 'scope' in reg.keys() and reg['scope'] == "packages":
+                packages = topology['packages']
+                threads_to_use = [packages[p][0] for p in packages]
+            elif 'scope' in reg.keys() and reg['scope'] == "cores":
+                cores = topology['cores']
+                threads_to_use = [cores[p][0] for p in cores]
+            else: # Default to threads
+                threads_to_use = range(self.helper.get_threads_count())
+            if len(reg_values) == len(threads_to_use):
+                value = 0
+                for t in threads_to_use:
+                    self.write_register(reg_name, reg_values[value], t)
+                    value += 1
+                ret = True
+        elif rtype in [RegisterType.MMCFG, RegisterType.PCICFG, RegisterType.MMIO] and bus_data:
+            values = len(bus_data)
+            if len(reg_values) == values:
+                for index in range(values):
+                    self.write_register(reg_name, reg_values[index], cpu_thread, bus_data[index])
+                ret = True
         else:
-            return False
+            if len(reg_values) == 1:
+                self.write_register(reg_name, reg_values[0])
+                ret = True
+        return ret
 
     def write_register_all_single(self, reg_name, reg_value, cpu_thread=0):
-        bus_data = self.get_register_bus( reg_name )
-        if not bus_data:
-            return False
-        values = len(bus_data)
-        for index in range( values ):
-            self.write_register( reg_name, reg_value, cpu_thread, index )
+        reg = self.get_register_def(reg_name)
+        rtype = reg['type']
+        bus_data = self.get_register_bus(reg_name)
+        if RegisterType.MSR == rtype:
+            topology = self.cpu.get_cpu_topology()
+            if 'scope' in reg.keys() and reg['scope'] == "packages":
+                packages = topology['packages']
+                threads_to_use = [packages[p][0] for p in packages]
+            elif 'scope' in reg.keys() and reg['scope'] == "cores":
+                cores = topology['cores']
+                threads_to_use = [cores[p][0] for p in cores]
+            else: # Default to threads
+                threads_to_use = range(self.helper.get_threads_count())
+            for t in threads_to_use:
+                self.write_register(reg_name, reg_value, t)
+        elif rtype in [RegisterType.MMCFG, RegisterType.PCICFG, RegisterType.MMIO] and bus_data:
+            for bus in bus_data:
+                self.write_register(reg_name, reg_value, cpu_thread, bus)
+        else:
+            self.write_register(reg_name, reg_value)
```

Comment:
Back to the previous note, if we have the explicit method to write a single value to all instances of a register, I am questioning even more the usefulness of the write_register_all()